### PR TITLE
feat: add ability to swap automatically between native usdc and bridg…

### PR DESCRIPTION
**Pull request by kjr217**
…ed usdc on perp reactor

# Task: RYSK-138

## Description

- add ability to swap automatically between native usdc and bridged usdc to allow use of perp protocols that dont yet support native usdc

An issue was raised where a user could theoretically modify the GmxHedgingReactor's state by making their own increase/decreasePositionRequests and entering the GmxHedgingReactor's address as the callback address in the function parameters. The user's request would been to be executed while the reactor has its own position request pending, meaning that the user would have to make a request immediately before our reactor such that their request executed before the reactors (positions are executed on a FCFS basis at the GMX contract level under normal circumstances, and the Arbitrum sequencer sequences transactions on a FCFS basis), or the GMX keeper system must experience a delay such that the user manually executes their own tx after the 3 minute window allows them to override the keeper system, while the reactor's request is pending.

This was enabled through a block of code in the gmxPositionCallback() function which decremented pendingIncreaseCallback/pendingDecreaseCallback without checking that the callback came from a position execution that belonged to the reactor. To fix this, a new state mapping was made that tracks they unique keys of the reactor's pending positions, and all state changes are wrapped within a check to ensure that the position key in the callback matches a pending position key for the reactor.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update
- [ ] Quality of life improvement
- [ ] Package upgrades

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have NATSPEC'd any new functions
- [ ] If appropriate, I have properly tested my new functionality

### CODE REVIEW

File: packages/contracts/contracts/hedging/PerpHedgingReactorWithSwap.sol [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+	////////////////////////////////////////////
+
+	/// @notice function to deposit 1 wei of USDC into the margin account so that a margin account is made, cannot be
+	///         be called if an account already exists
+	function initialiseReactor() external {
+		(IERC20 collateral, uint256 collat) = clearingHouse.getAccountCollateralInfo(accountId, collateralId);
+		if (collat != 0) {
+			revert();
+		}
+		SafeTransferLib.safeTransferFrom(collateralAsset, msg.sender, address(this), 1);
```
Comment (JoshDAO) [inline packages/contracts/contracts/hedging/PerpHedgingReactorWithSwap.sol:150]: need transferring

File: packages/contracts/contracts/hedging/PerpHedgingReactorWithSwap.sol [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+			tokenIn: collateralAsset,
+			tokenOut: LPcollateralAsset,
+			fee: 100,
+			recipient: parentLiquidityPool,
+			deadline: block.timestamp,
+			amountIn: _amount,
+			amountOutMinimum: _amountOutMinimum,
+			sqrtPriceLimitX96: 0
+		});
+		uint256 amountOut = swapRouter.exactInputSingle(params);
```
Comment (JoshDAO) [inline packages/contracts/contracts/hedging/PerpHedgingReactorWithSwap.sol:487]: this function isnt sending anything back to the liquidity pool

File: packages/contracts/contracts/hedging/GmxHedgingReactorWithSwap.sol [hunk 3]
```diff
@@ -0,0 +1,1094 @@
...
+import "../libraries/SafeTransferLib.sol";
+
+import "../interfaces/ILiquidityPool.sol";
+import "../interfaces/IHedgingReactor.sol";
+import "../interfaces/IRouter.sol";
+import "../interfaces/IPositionRouter.sol";
+import "../interfaces/IReader.sol";
+import "../interfaces/IGmxVault.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "hardhat/console.sol";
```
Comment (kjr217) [inline packages/contracts/contracts/hedging/GmxHedgingReactorWithSwap.sol:21]: remove

File: packages/contracts/contracts/hedging/PerpHedgingReactorWithSwap.sol [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+		// if delta comes in negative then the pool must go long
+		// if delta comes in positive then the pool must go short
+		// the signs must be flipped when going into _changePosition
+		// make sure the caller is the vault
+		require(msg.sender == parentLiquidityPool, "!vault");
+		deltaChange = _changePosition(-_delta);
+		// record the delta change internally
+		internalDelta += deltaChange;
+	}
+	// TODO: transferOut function
```
Comment (kjr217) [inline packages/contracts/contracts/hedging/PerpHedgingReactorWithSwap.sol:167]: need to remove

**Review by JoshDAO — APPROVED** [review]

**Pull request merged**